### ISSUE TITLE
Proc#curry on lambda with optional param

### ISSUE THREAD

**Issue by tak1n** [jruby/jruby#3011]
I'm currently  getting some 2.2 compat done for rubinius and found this bug:

```
rubinius ➤ ruby -v                                                                                                                                                                                        git:2.2*
ruby 2.2.2p95 (2015-04-13 revision 50295) [x86_64-linux]
rubinius ➤ irb                                                                                                                                                                                            git:2.2*
irb(main):001:0> lambda {|a,*b|}.arity
=> -2
irb(main):002:0> lambda {|a,*b|}.curry(3)
=> #<Proc:0x007fdac5168710 (lambda)>
irb(main):003:0> lambda {|a,b=nil|}.arity
=> -2
irb(main):004:0> lambda {|a,b=nil|}.curry(3)
ArgumentError: wrong number of arguments (3 for 1..2)
        from (irb):4:in `curry'
        from (irb):4
        from /home/benny/.rubies/2.2.2/bin/irb:11:in `<main>'


rubinius ➤ bin/ruby -v                                                                                                                                                                                    git:2.2*
rubinius 2.5.5.c97 (2.1.0 4b922b0e 2015-06-02 3.5.0 JI) [x86_64-linux-gnu]
rubinius ➤ bin/irb                                                                                                                                                                                        git:2.2*
irb(main):001:0> lambda {|a,*b|}.arity
=> -2
irb(main):002:0> lambda {|a,*b|}.curry(3)
=> #<Proc:0x2da0 (lambda)>
irb(main):003:0> lambda {|a,b=nil|}.arity
=> -2
irb(main):004:0> lambda {|a,b=nil|}.curry(3)
=> #<Proc:0x2fc4 (lambda)>
```

Unfortunately this also occurs on jruby, therefore I thought I should report it:

```
rubinius ➤ ruby -v                                                                                                                                                                                        git:2.2*
jruby 9.0.0.0.pre2 (2.2.2) 2015-04-28 2755ae0 OpenJDK 64-Bit Server VM 25.45-b02 on 1.8.0_45-b14 +jit [linux-amd64]
rubinius ➤ irb                                                                                                                                                                                            git:2.2*
irb(main):001:0> lambda {|a,*b|}.arity
=> -2
irb(main):002:0> lambda {|a,*b|}.curry(3)
=> #<Proc:0x17d0685f@uri:classloader:/jruby/kernel/proc.rb:45 (lambda)>
irb(main):003:0> lambda {|a,b=nil|}.arity
=> -2
irb(main):004:0> lambda {|a,b=nil|}.curry(3)
=> #<Proc:0x3891771e@uri:classloader:/jruby/kernel/proc.rb:45 (lambda)>
irb(main):005:0> 
```

**Comment by tak1n** [comment]
Seems like rbx share similar codebases for `Proc#curry`:
https://github.com/rubinius/rubinius/blob/master/kernel/common/proc.rb#L69-L83
vs
https://github.com/jruby/jruby/blob/master/core/src/main/ruby/jruby/kernel/proc.rb#L3-L17

I will try to send a pr after fixing it at rbx.

**Comment by headius** [comment]
@tak1n Yes, we share an impl for this logic. Thanks for taking it on!

**Comment by headius** [comment]
@tak1n Any progress?

**Comment by tak1n** [comment]
@headius sry hadn't much time lately :(

**Comment by headius** [comment]
JRuby appears to work properly now, likely due to fixes leading up to f7d1444d8678031422c0be7ba20ce296b17194d8 which was included in 9.1.8.0. Marking as fixed as of that release.

If you run into problems in the future, please file a new issue.
